ir/src/constraints/list_comprehension.rs

```diff
@@ -185,9 +187,18 @@ fn parse_list_folding(
     i: usize,
 ) -> Result<Expression, SemanticError> {
     match lf_type {
-        ListFoldingType::Sum(lc) | ListFoldingType::Prod(lc) => {
-            let iterable_context = build_iterable_context(lc)?;
-            let list = unfold_lc(lc, symbol_table)?;
+        ListFoldingType::Sum(lf_value_type) | ListFoldingType::Prod(lf_value_type) => {
+            let list = build_list_from_list_folding_value(lf_value_type, symbol_table)?;
+            let iterable_context =
+                if let ListFoldingValueType::ListComprehension(lc) = lf_value_type {
+                    build_iterable_context(lc)?
+                } else {
+                    BTreeMap::new()
+                };
+            assert!(
+                !list.is_empty(),
+                "List on which list folding is applied is empty."
+            );
```

Comment:
I think maybe this should either return a `SemanticError` or be a `debug_assert`. What was your thinking here?